### PR TITLE
DAOS-10896: Eberroca pipeline API impl feature branch merge

### PR DESCRIPTION
**Pull request by eberroca**
PR to merge my private branch with the feature branch. After this, this will be the official branch for the PIPELINE API feature until it is merged with master.

### CODE REVIEW

File: src/pipeline/cli_pipeline.c
Comment (wangdi1) [inline src/pipeline/cli_pipeline.c:657]: If there is always single shard task in dc_pipe_line(), then probably we do not need create another shard task here?  or  it will eventually have multiple shards task here?

File: src/pipeline/srv_pipeline.c
Comment (wangdi1) [inline src/pipeline/srv_pipeline.c:799]: check pool map version first?

File: src/pipeline/srv_pipeline.c
Comment (wangdi1) [inline src/pipeline/srv_pipeline.c:359]: input check? probably should be done on client side before packing RPC?

**Review by wangdi1 — COMMENTED** [review]
Only check cli/srv_pipeline.c, some questions and comments there. Some of the implementation make me think this probably should go into object directory,  though probably in future improvements.

File: src/pipeline/cli_pipeline.c
Comment (eberroca) [reply src/pipeline/cli_pipeline.c:632]: @wangdi1 Should I change it to DER_NOTSUPPORTED? Your call. I am fine either way.

File: src/pipeline/cli_pipeline.c
Comment (eberroca) [reply src/pipeline/cli_pipeline.c:270]: @wangdi1 Well...... Object API is not fully factorized, and a lot of common functionality between object and pipeline has to be replicated because object API functions (1) do more things that what I needed, and (2) uses internal object data structures and variables that can't be accessed outside the object directory.

When we started development, there was some discussions about where to put the pipeline code. One option was to put all the functionality within the Object API (make pipeline acceleration a feature of the Object API, instead of a new API). I think we decided against that, but I don't remember why. Maybe @mchaarawi can shed more light about this. In any case, I basically went over the object api to see how the code should be done and more or less replicate functionality but with the peculiarities of the pipeline thing.

File: src/pipeline/cli_pipeline.c
Comment (eberroca) [reply src/pipeline/cli_pipeline.c:160]: Exactly, all that part (replication) is probably going to be the same as object. Either that part is abstracted away from the object API and made it common for all APIs, or pipeline should ultimately move to object API.

File: src/pipeline/cli_pipeline.c
Comment (eberroca) [reply src/pipeline/cli_pipeline.c:645]: @wangdi1 Not sure what I am doing there. I think I copy that from somewhere in the object API. Maybe code I say in mid' 2021. It is incredible it works :-)

File: src/pipeline/cli_pipeline.c
Comment (eberroca) [reply src/pipeline/cli_pipeline.c:657]: @wangdi1 I imagine that when EC objects get supported, it may  need to have multiple tasks. But maybe, even in that case, this is still an overkill (for regular objects). So yeah, not sure what the best course of action is. It works now, but maybe it is better to just call the function directly instead of creating a task there.

File: src/pipeline/srv_pipeline.c
Comment (eberroca) [reply src/pipeline/srv_pipeline.c:359]: @wangdi1 Good point. I can move it to the client.

Comment (eberroca) [reply src/pipeline/srv_pipeline.c:359]: @wangdi1 Actually, we should not, for security reasons. But we can check it twice, in client first and then on server to discard malicious users.

File: src/pipeline/cli_pipeline.c
Comment (eberroca) [reply src/pipeline/cli_pipeline.c]: I changed it

File: src/client/api/pipeline.c
Comment (eberroca) [reply src/client/api/pipeline.c:36]: @jolivier23 : "I presume this is on the critical path for the API"

Actually, I don't think this is on the critical path. This allocations are done while the user is building the pipeline object. Once built, the user submits the object to the pipeline_run() function to do the filtering/aggregation. The user can submit the same pipeline object multiple times until the filtering/aggregation is done.

What is missing here, is that the object still has to allocate other objects. There is a separation of allocation responsibilities which is not good. Ultimately, the idea is for the user to create a pipeline object submitting the filtering/aggregation to be done as a high level string, or add enough API calls so the user can build the entire pipeline without the need to call malloc once.

File: src/include/daos/rpc.h
Comment (eberroca) [reply src/include/daos/rpc.h:55]: Changed

File: src/include/daos_pipeline.h
Comment (eberroca) [reply src/include/daos_pipeline.h:27]: @jolivier23 This actually started as an enumerated type, but @johannlombardi argued against it. I don't remember the exact reason, but he mention that it was easier to extend using strings this way.

File: src/client/api/pipeline.c
Comment (mchaarawi) [reply src/client/api/pipeline.c:36]: performance wise, this and filter_add should not usually be on the critical path.
however they are on an API critical path, as a use is going to call filter_add a lot, but maybe not a lot of pipeline_add. realloc on every filter add might be a lot, but i think we can add a comment here that mentions Jeff's idea or have something like a fixed buffer on the stack where we can add parts etc, till it runs out then we can allocate more. just some ideas that we can incorporate as a TODO IMO.

File: src/client/api/pipeline.c
Comment (mchaarawi) [inline src/client/api/pipeline.c:101]: should not assert here and return DER_INVAL instead if argument is NULL.

File: src/client/api/pipeline.c
Comment (mchaarawi) [inline src/client/api/pipeline.c:103]: same (should not assert on user input)

File: src/client/api/pipeline.c
Comment (mchaarawi) [inline src/client/api/pipeline.c:105]: same (should not assert on user input)

File: src/client/api/pipeline.c
Comment (mchaarawi) [inline src/client/api/pipeline.c:113]: curious as to why init here on a free call?

File: src/pipeline/filter.c
Comment (mchaarawi) [inline src/pipeline/filter.c:291]: alignment

File: src/pipeline/filter.c
Comment (mchaarawi) [inline src/pipeline/filter.c:341]: on failure, comp_pipe->filters is leaked

File: src/pipeline/cli_pipeline.c
Comment (mchaarawi) [inline src/pipeline/cli_pipeline.c:50]: please add like a TODO comment to do a pool map refresh to update this

File: src/pipeline/cli_pipeline.c
Comment (mchaarawi) [reply src/pipeline/cli_pipeline.c:160]: it's better to keep separate for now to avoid a lot of merge conflict. i think once some picks this up again, we can merge into object module.

File: src/pipeline/cli_pipeline.c
Comment (mchaarawi) [inline src/pipeline/cli_pipeline.c:207]: hardcoded 20 sounds like something to be defined in a header

File: src/pipeline/filter_funcs.c
Comment (mchaarawi) [inline src/pipeline/filter_funcs.c:318]: no need to use goto here and can just return rc?

File: src/pipeline/aggr_funcs.c
Comment (eberroca) [reply src/pipeline/aggr_funcs.c:29]: ok

File: src/pipeline/aggr_funcs.c
Comment (eberroca) [reply src/pipeline/aggr_funcs.c:52]: ok

**Review by mchaarawi — CHANGES_REQUESTED** [review]

File: src/pipeline/common_pipeline.c
Comment (eberroca) [reply src/pipeline/common_pipeline.c:52]: done

File: src/pipeline/getdata_funcs.c
Comment (eberroca) [reply src/pipeline/getdata_funcs.c:26]: done

File: src/client/api/pipeline.c
Comment (eberroca) [reply src/client/api/pipeline.c:36]: @mchaarawi OK. I added a "TODO" comment. Future developer should change this.

File: src/client/api/pipeline.c
Comment (eberroca) [reply src/client/api/pipeline.c:101]: ok

File: src/client/api/pipeline.c
Comment (eberroca) [reply src/client/api/pipeline.c:103]: ok

File: src/client/api/pipeline.c
Comment (eberroca) [reply src/client/api/pipeline.c:105]: ok

File: src/pipeline/filter.c
Comment (eberroca) [reply src/pipeline/filter.c:318]: Actually, yes. I have to change the exit part of the code

File: src/client/api/pipeline.c
Comment (mchaarawi) [reply src/client/api/pipeline.c:36]: sorry i should have added, as long as it's OK with Jeff. addressing this before landing would be preferable, but if Jeff is ok with it, we can leave it as todo :-)

File: src/pipeline/filter.c
Comment (eberroca) [reply src/pipeline/filter.c:341]: fixed

File: src/pipeline/filter.c
Comment (eberroca) [reply src/pipeline/filter.c:345]: fixed

File: src/pipeline/cli_pipeline.c
Comment (eberroca) [reply src/pipeline/cli_pipeline.c:50]: done

File: src/pipeline/filter_funcs.c
Comment (eberroca) [reply src/pipeline/filter_funcs.c:318]: right

File: src/pipeline/filter.c
Comment (eberroca) [reply src/pipeline/filter.c:131]: I am not so sure how to do that, so I will leave a TODO comment

Comment (eberroca) [reply src/pipeline/filter.c:131]: Actually, I can try to do it that way, if it is a must for the landing..

**Pull request merged**
